[feature/#2] [홈탭] 스폰서 자동 스크롤

l2hyunwoo opened the pull request:
## Issue
- close #2 

## Overview (Required)
- ~기능 개발을 하다가 궁금한 점들이 있어, 코드리뷰를 받고자 먼저 PR을 날리게 되었습니다.~
- ~Question~
     - ~1. BindingAdapter 내에서 suspend 함수를 사용하고자 CoroutineScope이 필요했는데 Dispatchers.Main.immediate을 사용하기에는 Fragment의 View가 destroy 되어도 함수가 계속 작동할 수 있을 것 같아 정상적인 작동이 보장되지 않을 것 같았습니다. 그래서  ``findViewTreeLifecycleOwner()?.lifecycleScope``을 사용하려 했으나 ``findViewTreeLifecycleOwner``가 null이 나오더군요. Fragment의 뷰가 다 만들졌는데 RecyclerView의 ViewTreeLifecycleOwner가 Null이 뜨는 이유가 궁금합니다. 아래 코드를 참고를 했는데도 불구하고 원인이 이해가 안되었습니다. 제가 이해하기로는 View의 parent를 탐색하여 Null이 아닌 LifecycleOwner를 찾는 것이었습니다.~
     - ~2. 위의 문제로 인해 Fragment의 LifecycleScope을 바인딩 어댑터 부분에 주입을 하였는데, Fragment의 ``onDestroyView``가 호출된 이후에는 LifecycleScope가 null이 될텐데 이 때문에 NullPointerExcpetion이 발생될 수 있을 것 같은데 제 예상이 맞은가요?~
     - ~3. 이슈에 주신 링크를 보았는데 ComposeView를 기존 패턴에서 사용하기 위해서는 onCreateViewHolder와 onBindViewHolder를 모두 사용해야할 것 같은데, ListBindingAdapter 대신 ComposeAdapter, ComposeViewHolder와 같은 Util Adapter를 제작해서 사용하는 건 어떤지 의견을 듣고 싶습니다.~

```kotlin
    @Nullable
    public static LifecycleOwner get(@NonNull View view) {
        LifecycleOwner found = (LifecycleOwner) view.getTag(R.id.view_tree_lifecycle_owner);
        if (found != null) return found;
        ViewParent parent = view.getParent();
        while (found == null && parent instanceof View) {
            final View parentView = (View) parent;
            found = (LifecycleOwner) parentView.getTag(R.id.view_tree_lifecycle_owner);
            parent = parentView.getParent();
        }
        return found;
    }
```

~**Updated 21/08/28**~
- ~아래의 이슈만 해결이 된다면 머지를 할 수 있을 것 같습니다.~
~Q) ``NetworkImage`` 클래스에서 rememberImagePainter를 사용하는데 이미지가 로드가 되지 않고 painter.state가 항상 empty가 되는 현상을 발견했습니다. 로컬에서 이미지를 받아올 때에는 이미지 로드가 잘 되는 것을 확인을 했었습니다.~
- ~로깅을 해봤을 때 painter.request의 dispatcher가 Dispatchers.IO로 되어있어서~
```kotlin
internal fun NetworkImage(
    imageUrl: String,
    nonSuccessTintColor: Color,
    modifier: Modifier = Modifier,
    builder: ImageRequest.Builder.() -> Unit = {
        placeholder(R.drawable.ic_android_white)
        error(R.drawable.ic_android_white)
        dispatcher(Dispatchers.Main.Immediate)
    },
)
```
~로도 변경을 해보았으나 이 역시 실행이 안되었습니다.~
~디버그를 해도 감이 잡히지 않아서 질문드립니다 🙏🏻 @Pluu~

## Links
- ComposeAdapter 참고자료:
https://github.com/android/sunflower/blob/compose_recyclerview/app/src/main/java/com/google/samples/apps/sunflower/adapters/PlantAdapter.kt

## Screenshot **2021.09.26 UPDATED**
Before | After
:--: | :--:
<img src="https://user-images.githubusercontent.com/54518925/128627847-482c03fe-9e33-465d-bd81-c955692c5985.png" width="300" /> | <video src="https://user-images.githubusercontent.com/54518925/134800337-40052975-b25a-4e9c-99b8-98c0bbe3cfb8.mp4" width="300" />

l2hyunwoo committed:
[feature/#2] 스폰서 무한 스크롤 기능 구현

l2hyunwoo committed:
[feature/#2] Compose gradle config

l2hyunwoo committed:
[feature/#2] SponsorItem 만들기

Pluu commented:
@l2hyunwoo 3번 질문에 대해서는
외부에서 사용하는 "RecyclerView.Adapter 및 ViewHolder 패턴을 유지"가 조건이므로 컴포즈용으로 추가해도 됩니다.

Pluu commented:
@l2hyunwoo 로직을 기준으로 실제 디버깅을 해보시면 Fragment는 LifecycleOwner를 실제로 가지고 있지만, ItemView는 가져왔을대 비어있는것을 알 수 있습니다.

![image](https://user-images.githubusercontent.com/1534926/128634809-87ed75cd-1116-4ade-ac3b-f2f1a8aa8fce.png)

```java
ViewParent parent = view.getParent();
while (found == null && parent instanceof View) {
```
getParent()를 통해서 뷰를 찾아야하는데 parent가 null이므로 LifecycleOwner도 null로 반환됩니다.
Adapter에서 사용을 해야한다면 parent 혹은 recyclerview를 통해서 가져오는 방법을 사용하시면 됩니다.

l2hyunwoo committed:
[feature/#2] LifecycleScope가 null일 때에 대응

l2hyunwoo committed:
[feature/#2] Compose Migration

- issue: 로컬 이미지는 paint가 되는데 network를 통해서 받아온 이미지에 대해서 load가 안됨

l2hyunwoo commented:
@Pluu 오랜만에 업데이트를 드렸습니다! 이슈와 관련된 질문을 남겼습니다 확인해주시면 정말로 감사하겠습니다!!

Pluu commented:
@l2hyunwoo 문의하신 내용 저도 체크해보겠습니다

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/SponsorAdapter.kt:
```diff
@@ -1,34 +1,95 @@
+    itemHandler: SponsorAdapter.ItemHandler
+) {
+    Column(
+        Modifier.wrapContentHeight()
```
@l2hyunwoo 63번라인과 69번라인에서 `wrapContentHeight`과 `wrapContentWidth` 작성으로 Coil State가 Empty로만 나오는것 같네요.

l2hyunwoo committed:
[feature/#2] Compose Migration

- 이미지 스크롤 완료

l2hyunwoo reviewed (commented):
@Pluu 

근 3주간 바빠서 제대로 코드를 보지 못했습니다. 플루님 말씀처럼 코드 한 줄만 바꾸면 금방 고쳐졌습니다.(고친 후의 영상은 After에 업데이트 했습니다!)

문제 해결은 했지만 코드 내부를 봐도 Modifier에 wrapContent를 줄 때 이미지 로드를 할 시 문제가 발생하는 원인이 파악되지가 않아서 혹시 어떤 부분에서 그렇게 원인을 추정하셨는 지 알 수 있을까요?

l2hyunwoo committed:
Merge branch 'main' of https://github.com/l2hyunwoo/DroidKnights2021_App into feature/#2

l2hyunwoo committed:
Merge branch 'main' of https://github.com/droidKnights/DroidKnights2021_App into feature/#2

l2hyunwoo committed:
[feature/#2] Conflict resolve

l2hyunwoo committed:
[feature/#2] 안 쓰는 코드 삭제

Pluu commented:
@l2hyunwoo onBindViewHolder에서 발생하는 LifecycleOwner이슈는 아래글을 참고해주세요.
http://pluu.github.io/blog/android/2021/09/20/lifecycleowner/

Pluu commented:
@l2hyunwoo 해당 이슈의 원인은 찾지못했습니다.

`After`으로 올려주신 영상을 확인했는데, 왼쪽끝에 도달할때 이미지의 동작이 다시 오른쪽으로 밀리는 현상이 보이는데 해당 부분의 수정을 부탁드려도 될까요?

l2hyunwoo commented:
> After으로 올려주신 영상을 확인했는데, 왼쪽끝에 도달할때 이미지의 동작이 다시 오른쪽으로 밀리는 현상이 보이는데 해당 부분의 수정을 부탁드려도 될까요?

최종적으로 수정 후 다시 태그해드리겠습니다! 위의 말씀해주신 부분을 제외하고 다 해결되었습니다.

l2hyunwoo committed:
[feature/#2] Horizontal 스크롤 시 발생하던 이슈 해결

l2hyunwoo commented:
수정되었고 영상 업데이트했습니다 확인 부탁드립니다 🙏🏻 @Pluu

Pluu reviewed (commented):
@fornewid Coil 사용시 wrap으로 하면 상태가 항상 empty인데 뭔가 아시나요?

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -31,12 +45,34 @@ internal class InfoAdapter(
+        }
+
+        coroutineScope.launch {
+            holder.binding.sponsorList.launchAutoScroll()
```
@l2hyunwoo 일반적으로 AutoScroll시에 적용하는 사용자 터치에 의한 일시 중지 처리로직이 누락되어 있네요.

그래서 AutoScroll 중에 터치와 스크롤 동작이 중복으로 발생하고 있습니다.

l2hyunwoo replied:
아 이 기능을 처음 구현해봐서 이런 로직이 필요한지는 처음 알게되었습니다! 확인 후 수정해보도록 하겠습니다.

l2hyunwoo replied:
@Pluu 해당 기능 구현했습니다! 확인 부탁드립니다 🙏🏻

Pluu replied:
@l2hyunwoo QA 중에 있다는게 어떤 의미일까요?

l2hyunwoo replied:
아 잘못 올라간겁니다! 댓글 지웠어야 했는데 작성한 것 자체를 잊었습니다.

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -31,12 +45,34 @@ internal class InfoAdapter(
+            val currentList = (adapter as SponsorAdapter).currentList
+            val secondPart = currentList.subList(0, firstVisibleItem)
+            val firstPart = currentList.subList(firstVisibleItem, currentList.size)
+            (adapter as SponsorAdapter).submitList(firstPart + secondPart)
```
@l2hyunwoo 사용자 터치에 의한 제어가 방어되어있지 않아서 아래와같은 현상도 재현되네요.
List를 강제로 합칠때 current/seccond/firstPart 불일치로 발생할 것으로 보여요 

https://user-images.githubusercontent.com/1534926/134142641-7feff4c8-439d-4f3e-bb0e-9f628c8d6ec3.mp4

Pluu replied:
@l2hyunwoo DX 수치를 높게하면 submitList 계산으로 동일한 현상을 볼 수 있네요.

Pluu reviewed (changes requested):
@l2hyunwoo AutoScroll 동작으로 인해서 오류가 발생하는 케이스가 있네요.

fornewid commented:
> @fornewid Coil 사용시 wrap으로 하면 상태가 항상 empty인데 뭔가 아시나요?

compose에서 wrap을 쓰면 크기를 알 수 없어서 처리가 안될걸요?

의도된 동작으로 알고 있습니다.
https://github.com/coil-kt/coil/issues/862

l2hyunwoo committed:
[feature/#2] 스크롤 시 뷰에 터치를 하면 자동 스크롤 멈추게 하기!

l2hyunwoo committed:
[feature/#2] 스크롤 시 뷰에 터치를 하면 자동 스크롤 멈추게 하기!

l2hyunwoo commented:
@Pluu 추가 기능구현 및 영상 업데이트 되었습니다

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -31,12 +42,66 @@ internal class InfoAdapter(
+            }
+        })
+
+        holder.binding.sponsorList.setOnTouchListener { _, event ->
```
@l2hyunwoo Lint warning이 노출되고 있으므로 확인 부탁해요

l2hyunwoo replied:
@Pluu setOnTouchListener를 사용할 때 사용자 접근성(장님이나 시력이 안 좋으신 분들)에게 편의를 주고자 performClick을 뷰 내에서 추가적으로 구현하는 것을 권장하기 위해 lint를 건 것 같습니다. 이 부분 ``viewBindViewHolder``에 ``@SupperssLint("ClickableViewAccessibility")``어노테이션을 붙여서 해결하고자 합니다.

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -31,12 +42,66 @@ internal class InfoAdapter(
+                    true
+                }
+                MotionEvent.ACTION_UP -> {
+                    scrollJob = coroutineScope.launch {
```
@l2hyunwoo scrollJob의 초기화와 cancel이 2곳에서 적용되고 있네요. 중복 코드이며 복수의 곳에서 진행시 사이드 이펙트의 위험이 있습니다. 확인 부탁해요

l2hyunwoo replied:
scrollJob을 InfoAdapter의 멤버변수로 놓겠습니다!

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -31,12 +42,66 @@ internal class InfoAdapter(
+            holder.binding.sponsorList.launchAutoScroll()
+        }
+
+        holder.binding.sponsorList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
```
@l2hyunwoo 화면 높이가 좁은 경우, holder.binding.sponsorList가 재활용되므로 복수개의 스크롤 리스너가 추가되어 각 아이템의 폭이 점점 넓어지는 현상이 일어납니다.

l2hyunwoo replied:
이 함수를 제거해도 동일하게 동작되네요(위의 로직과 중복이기도 하네요). setOnTouchListener만 살려놓겠습니다.

Pluu reviewed (changes requested):
이슈가 되는 부분을 다시 추가했습니다. 🙏🤦‍♂️

l2hyunwoo committed:
[feature/#2] 코드리뷰 반영

- 멤버변수로 job을 설정하여 중복초기화 문제 해결
- 중복로직 삭제
- async, await을 활용하여 무조건 첫번째 element에서 계산되게 적용

l2hyunwoo committed:
[feature/#2] 코드리뷰 반영

- 스크롤 추가 수정 (SCROLL_DX 5만으로 주고 2분간 테스트 했으나 이슈 발생하지 않음)

l2hyunwoo commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -21,6 +30,8 @@ internal class InfoAdapter(
         }
     }
 
+    private var scrollJob: Job? = null
+
```
scrollJob을 멤버변수로 빼냈습니다!

l2hyunwoo commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -29,14 +40,66 @@ internal class InfoAdapter(
+            holder.binding.sponsorList.launchAutoScroll(coroutineScope)
+        }
+
+        holder.binding.sponsorList.setOnTouchListener { _, event ->
```
setOnTouchListener만 남겨두고 scrollListener는 지웠습니다 (중복로직)

l2hyunwoo commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -29,14 +40,66 @@ internal class InfoAdapter(
         return R.layout.item_info_header
     }
 
+    @SuppressLint("ClickableViewAccessibility")
```
접근성 문제로 인해 RecyclerView를 다시 커스텀해서 performClick을 구현하거나 SuppressLint 어노테이션을 붙여야 해서 후자를 선택했습니다.

l2hyunwoo commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -29,14 +40,66 @@ internal class InfoAdapter(
+            (layoutManager as LinearLayoutManager).findFirstVisibleItemPosition()
+        if (firstVisibleItem != RecyclerView.NO_POSITION && firstVisibleItem != 0) {
+            coroutineScope.launch {
+                val defferedAutoScroll = async {
+                    val currentList = (adapter as SponsorAdapter).currentList.toMutableList()
+                    val secondPart = currentList.subList(0, REPLACE_POSITION)
+                    val firstPart = currentList.subList(REPLACE_POSITION, currentList.size)
+                    val newList = mutableListOf<Sponsor>().apply {
+                        addAll(firstPart)
+                        addAll(secondPart)
+                    }
+                    (adapter as SponsorAdapter).submitList(newList)
+                }
+                defferedAutoScroll.await()
```
SCROLL_DX = 50000을 주고 2분동안 앱이 실행되는 동안 문제가 발생하지 않았습니다.

- plus의 퍼포먼스 문제가 있는 것 같아 emptyList에 addAll을 해주는 방식으로 로직을 변경했습니다.
- firstVisibleItem이 3이 되는 경우에 스크롤 문제가 발생을 하여 리스트가 교체되는 동안 스크롤이 되어서 문제가 발생하지 않았나는생각을 가지게 되어 교체로직을 async, await로 기다리고 이후에 스크롤을 하는 것으로 로직을 변경했습니다.

혹시 제 로직에 문제점이 있다면 피드백 부탁드립니다 🙏🏻

l2hyunwoo reviewed (commented):
@Pluu 코드리뷰 반영했습니다!

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -29,14 +40,66 @@ internal class InfoAdapter(
         fun clickSponsor(sponsor: Sponsor)
     }
+
+    private tailrec suspend fun RecyclerView.launchAutoScroll(coroutineScope: CoroutineScope) {
```
@l2hyunwoo 해당 함수가 화면에 노출되지않는 경우에도 무한루프가 도는것을 확인했어요.

1. 홈키로 앱이 백그라운드 상태가 된 경우
2. 메인화면에 다른 탭을 누른 경우
3. 다른 Activity가 호출된 경우

- - - 
위 케이스로 봤을때 LifecycleOwner를 사용하지않아서 발생하는 이슈로 보이네요.

Pluu reviewed (changes requested):
스크롤 자체는 움직이는것은 더이상 현상은 안보였고, 무한루프 이슈가 확인되어서 리뷰남깁니다.

l2hyunwoo committed:
[feature/#2] 코드리뷰 반영

- LifeCycleObserver 활용해서 뷰가 파괴되지 않는 ON_STOP에서도 스크롤이 중단될 수 있게 설정

l2hyunwoo committed:
[feature/#2] 코드리뷰 반영

- scrollJob을 null 여부로 판단하는 대신에 isCancelled로 판단

l2hyunwoo commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -25,18 +37,111 @@ internal class InfoAdapter(
         submitList(listOf(InfoItem(sponsors)))
     }
 
+    class SponsorViewLifeCycleObserver(
+        private val holder: DataBindingViewHolder<InfoItem>
+    ) : LifecycleObserver {
```
LifecycleObserver로 ViewTreeLifeCycle에 맞춰서 scrollJob을 수행시키는 방향으로 로직을 수정했습니다.

l2hyunwoo commented on features/home/src/main/java/com/droidknights/app2021/home/ui/adapter/InfoAdapter.kt:
```diff
@@ -25,18 +37,111 @@ internal class InfoAdapter(
+        @SuppressLint("ClickableViewAccessibility")
+        @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+        fun stopScroll() {
+            (holder.binding as ItemInfoHeaderBinding).sponsorList.setOnTouchListener(null)
```
setOnTouchListener가 화면에 다시 돌아올 때마다 중복 set되는 것을 막기 위해 ON_STOP 시 null로 설정하게했습니다.

l2hyunwoo reviewed (commented):
@Pluu 생명주기에 유의하여 로직을 수정해봤습니다!!

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/HomeFragment.kt:
```diff
@@ -50,4 +50,9 @@ class HomeFragment : Fragment() {
     }
+
+    override fun onDestroyView() {
+        _binding = null
```
@l2hyunwoo DataBinding패턴에서 해당 로직을 사용한 이유를 잘 모르겠네요.

해당 패턴은 ViewBinding에서 사용하는 형태인데, 어떤 이유로 명시적 null을 했는지 설명이 필요해보이네요

Pluu commented on features/home/src/main/java/com/droidknights/app2021/home/ui/HomeFragment.kt:
```diff
@@ -7,31 +7,31 @@ import android.view.ViewGroup
     private val viewModel by viewModels<HomeViewModel>()
 
-    private lateinit var binding: FragmentHomeBinding
+    private var _binding: FragmentHomeBinding? = null
```
@l2hyunwoo 해당 패턴은 ViewBinding을 위한 패턴입니다.

`HomeFragment`는 DataBinding을 유지해야합니다

Pluu reviewed (changes requested):
기존 바인딩 로직을 변경된 부분이 발생했네요.
코드 수정 범위가 너무 넓어지지 않도록 처리를 해주시면 감사해요.

l2hyunwoo committed:
[feature/#2] binding 코드 원상복구

Pluu reviewed (approved):
긴 시간동안 수고하셨습니다.
